xwiki-commons-core/xwiki-commons-job/src/main/java/org/xwiki/job/internal/DefaultGroupedJobInitializerManager.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.job.internal;
+
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.slf4j.Logger;
+import org.xwiki.cache.Cache;
+import org.xwiki.cache.CacheException;
+import org.xwiki.cache.CacheManager;
+import org.xwiki.cache.config.LRUCacheConfiguration;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.manager.ComponentLookupException;
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.component.phase.Initializable;
+import org.xwiki.component.phase.InitializationException;
+import org.xwiki.job.GroupedJobInitializer;
+import org.xwiki.job.GroupedJobInitializerManager;
+import org.xwiki.job.JobGroupPath;
+import org.xwiki.job.JobManagerConfiguration;
+
+/**
+ * A component dedicated to find the appropriate {@link GroupedJobInitializer} based on a {@link JobGroupPath}.
+ *
+ * @since 12.5RC1
+ * @version $Id$
+ */
+@Component
+@Singleton
+public class DefaultGroupedJobInitializerManager implements GroupedJobInitializerManager, Initializable
+{
+    @Inject
+    @Named("context")
+    private Provider<ComponentManager> componentManagerProvider;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private GroupedJobInitializer defaultGoupedJobInitializer;
+
+    @Inject
+    private CacheManager cacheManager;
+
+    @Inject
+    private JobManagerConfiguration configuration;
+
+    private Cache<GroupedJobInitializer> cachedInitializers;
```

Comment:
Should maybe flush that cache when a GroupedJobInitializer component is registered/unregistered (ComponentDescriptorAddedEvent/ComponentDescriptorRemovedEvent).

Reply:
I hesitated about the way to deal with that cache: right now, if you check on #search method I check first on the cache and then build back it when I iterate over the components. My first move was to build the cache once and to rebuilt it in case of component added/removed but not sure how to ensure that the cache is big enough: what happens if I only rely on the cache and there's 101+ components? Some of them will never be cached and so I won't be able to reach them. Right now the strategy is more lazy: I put info in the cache as I need them and I never rely entirely on it, so I don't think there's a real need for listening on events.

Reply:
Lazy is the way to go but the problem is that you might rely on a GroupedJobInitializer instance which does not work anymore because it was unloaded for example.

Reply:
Right, so I need to flush the cache in case of removed event only then.

Reply:
I would still clean on both ComponentDescriptorAddedEvent and ComponentDescriptorRemovedEvent.

Reply:
Can do but I don't really see the point for cleaning on added event: we'd lose some values that are still correct in the cache...

Reply:
You can imagine installing a new GroupedJobInitializer being associated with a path which is already in the cache because it was inherited from some parent because there was no GroupedJobInitializer back then for this specific path.

Reply:
> a new GroupedJobInitializer being associated with a path which is already in the cache because it was inherited from some parent because there was no GroupedJobInitializer back then for this specific path.

I only keep the exact JobGroupPath mapping in cache (the value given by the GroupedJobInitializer#getId), so that scenario is not possible.

Reply:
So I did it only for remove event, it's easy to improve later if we want. See: https://github.com/xwiki/xwiki-commons/pull/96/files#diff-7024b53c2653ba2d321478a35653d515R156

Reply:
> I only keep the exact JobGroupPath mapping in cache (the value given by the GroupedJobInitializer#getId), so that scenario is not possible.

I was expecting inherited values to be put in the cache too to to make search faster.

Reply:
> I was expecting inherited values to be put in the cache too to make search faster

The rationale was actually that you won't call twice the initializer with the same jobGroupPath so it felt useless, but I'm realizing while writing this that it's actually the manager and not the initializer here and you'll definitely call it at least twice to get the initializer (once for executor and once for the JobGroupPathLockTree), so my argument is not really good anymore...